caravel/tasks.py

```diff
@@ -0,0 +1,160 @@
+    :param sql: sql query that will be executed
+    :param user_id: integer
+    :param milis: integer, unit time in miliseconds
+    :return: models.Query
+    """
+
+    query = models.Query()
+    query.user_id = user_id
+    query.database_id = database_id
+    if app.config.get('SQL_MAX_ROW'):
+        query.query_limit = app.config.get('SQL_MAX_ROW')
+    query.query_name = str(milis)
+    query.query_text = sql
+    # TODO(bkyryliuk): run explain query to derive the tables and fill in the
+    #                  table_ids
+    # TODO(bkyryliuk): check the user permissions
+    query.start_time = milis
+    query.query_status = models.QueryStatus.IN_PROGRESS
+    return query
+
+
+def is_query_select(sql):
+    return sqlparse.parse(sql)[0].get_type() == 'SELECT'
+
+
+# if sqlparse provides the stream of tokens but don't provide the API
+# to access the table names, more on it:
+# https://groups.google.com/forum/#!topic/sqlparse/sL2aAi6dSJU
+# https://github.com/andialbrecht/sqlparse/blob/master/examples/
+# extract_table_names.py
+#
+# Another approach would be to run the EXPLAIN on the sql statement:
+# https://prestodb.io/docs/current/sql/explain.html
+# https://cwiki.apache.org/confluence/display/Hive/LanguageManual+Explain
+def get_tables(sql):
+    # TODO(bkyryliuk): implement
+    return []
+
+
+def create_table_as(sql, table_name):
+    """Reformats the query into the create table as query.
+
+    :param sql: string, sql query that will be executed
+    :param table_name: string, will contain the results of the query execution
+    :return: string, create table as query
+    """
+    # TODO(bkyryliuk): split into atomic sql queries
+    # TODO(bkyryliuk): drop table if allowed, check the namespace and
+    #                  the permissions.
+
+    sql_statements = sqlparse.split(sql)
+    updated_sql = ""
+    for sql_statement in sql_statements:
+        if not sql_statement:
+            continue
+        if is_query_select(sql):
+            sql_statement = "CREATE TABLE IF NOT EXISTS %s AS %s" % (
+                table_name, sql_statement)
+        updated_sql += sql_statement + "\n"
+    return updated_sql
+
+
+def get_session():
+    """Creates new SQLAlchemy scoped_session."""
+    engine = create_engine(
+        app.config.get('SQLALCHEMY_DATABASE_URI'), convert_unicode=True)
+    return scoped_session(sessionmaker(
+        autocommit=False, autoflush=False, bind=engine))
+
+
+@celery_app.task
+def get_sql_results(database_id, sql, user_id):
+    """Executes the sql query returns the results.
+
+    :param database_id: integer
+    :param sql: string, query that will be executed
+    :param user_id: integer
+    :return: dataframe, query result
+    """
+    # Create a separate session, reusing the db.session leads to the
+    # concurrency issues.
+    query_session = get_session()
+    try:
+        db_to_query = query_session.query(models.Database).filter_by(
+            id=database_id).first()
```

Comment:
.first() doesn't fail if there's nothing, it only passes `None`, no need for exception handling here I think.